text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+- Feature Name: n/a
+- Start Date: 2015-10-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC describes how we intend to modify the compiler to support IDEs. The
+intention is that support will be as generic as possible. A follow-up internals
+post will describe how we intend to focus our energies and deploy Rust support
+in actual IDEs.
+
+There are two sets of technical changes proposed in this RFC: changes to how we
+compile, and the creation of an 'oracle' tool (name of tool TBC).
+
+This RFC is fairly detailed, it is intended as a straw-man plan to guide early
+implementation, rather than as a strict blueprint.
+
+
+## Compilation model
+
+An IDE will perform two kinds of compilation - an incremental check as the user
+types (used to provide error and code completion information) and a full build.
+The full build is explicitly signaled by the user (it could also happen
+implicitly, for example when the user saves a file). A full build is basically
+just a `cargo build` command, as would be done from the command line. It will
+take advantage of any future improvements to regular compilation (such as
+incremental compilation), but there is essentially no change from a compile
+today. It is not very interesting and won't be discussed further.
+
+The incremental check follows a new model of compilation. This check must be as
+fast as possible but does not need to generate machine code. We'll describe it
+in more detail below. We call this kind of compilation a 'quick-check'.
+
+This RFC also covers making compilation more robust.
+
+
+## The oracle
+
+The oracle is a long running daemon process. It will keep a database
+representation of an entire project's source code and semantic information (as
+opposed to the compiler which operates on a crate at a time). It is
+incrementally updated by the compiler and provides an IPC API for providing
```

Comment:
The major reason is parallelism - we want the IDE to carry on doing its thing whilst the compiler (coordinated by the oracle) compiles the non-urgent parts of changed source. Then the oracle needs to update its database with the results. Although the IDE could handle the threading to make all this work, it is simpler (and more reusable) just to do it all in a separate process.

With a decent IPC library, having an IPC API is not a lot more complex than having an FFI API. And in turns of implementation having a totally separate process is marginally easier.

In terms of stability, I don't think there is much difference between the two. An IPC API might be a little more robust because using an intermediate data structure and having parsing/serialisation adds a little abstraction.